---
params:
  sample: Sample
  normalized_sce: "data/anderson-single-cell/normalized/normalized_GSM4186961_sce.rds"
  marker_genes: "marker-genes/mapped_marker_genes.tsv"
  date: !r Sys.Date()
  gene_sets_present: FALSE
  ensembl_id_column: "ensembl"
  gene_symbol_column: NULL

title: "`r glue::glue('Gene expression plots for {params$sample}')`"
author: "CCDL"
date: "`r params$date`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    code_folding: hide
---

## Set Up

```{r, setup, include=FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE
)
```

```{r}
# Load libraries
library(magrittr)
library(readr)
library(SingleCellExperiment)
library(scater)
library(ggplot2)
library(ComplexHeatmap)

# Source custom functions script
source(file.path("utils", "marker-genes-analysis-functions.R"))
```

## Read in data

```{r}
normalized_sce <- read_rds(file.path(params$normalized_sce))

marker_genes <- read_tsv(file.path(params$marker_genes))

num_cells <- dim(normalized_sce)[2]
glue::glue("This sample has {num_cells} cells.")
```

## Perform Hierarchical Clustering

```{r}
ensembl <- params$ensembl_id_column

# get logcounts from normalized sce object
normalized_sce_logcounts_matrix <- as.matrix(t(logcounts(normalized_sce)))

# filter counts matrix to only data associated with the provided marker genes
normalized_sce_logcounts_matrix <- normalized_sce_logcounts_matrix[,colnames(normalized_sce_logcounts_matrix) %in% marker_genes$ensembl]

# transform counts into z-scores
normalized_sce_zscores_matrix <- scale(normalized_sce_logcounts_matrix,
                                    center = TRUE, scale = TRUE)

if (!is.null(params$gene_symbol_column)) {
  gene_symbol <- params$gene_symbol_column
  # we will want to replace the column names our matrix with the relevant gene
  # symbols for plotting
  map_ensembl_symbols <-
    data.frame(ensembl = colnames(normalized_sce_zscores_matrix)) %>%
    dplyr::left_join(marker_genes, by = "ensembl")
  
  colnames(normalized_sce_zscores_matrix) <-
    map_ensembl_symbols$gene_symbol
  
  # create column annotation data frame
  annotation_df <- marker_genes %>%
    dplyr::select(gene_symbol, gene_set) %>%
    dplyr::distinct()
  
  rownames(annotation_df) <- NULL
  
  # filter to ensure only those genes of that marker genes that had data stored
  # in the sce object are kept in the annotation object
  annotation_df <- annotation_df %>%
    dplyr::filter(gene_symbol %in% colnames(normalized_sce_zscores_matrix)) %>%
    tibble::column_to_rownames("gene_symbol")
  
  if (params$gene_sets_present == TRUE) {
    gene_set_names <- annotation_df %>%
      dplyr::arrange(gene_set) %>%
      dplyr::pull(gene_set) %>%
      unique()
    
    # get colors for the annotation object
    annotation_colors <-
      palette.colors(palette = "Okabe-Ito")[2:(length(gene_set_names) + 1)]
    
    gene_set_colors = annotation_colors[1:length(gene_set_names)]
    names(gene_set_colors) = gene_set_names
    gene_set_colors <- as.list(gene_set_colors)
    
    # create the column annotation for the ComplexHeatmap
    column_annotation <- HeatmapAnnotation(
      df = annotation_df,
      col = gene_set_colors,
      annotation_label = c("Gene Set")
    )
  } else {
    column_annotation <- NULL
  }
} else if (is.null(params$gene_symbol_column)) {
    # create column annotation data frame
  annotation_df <- marker_genes %>%
    dplyr::select(ensembl, gene_set) %>%
    dplyr::distinct()
  
  rownames(annotation_df) <- NULL
  
  # filter to ensure only those genes of that marker genes that had data stored
  # in the sce object are kept in the annotation object
  annotation_df <- annotation_df %>%
    dplyr::filter(ensembl %in% colnames(normalized_sce_zscores_matrix)) %>%
    tibble::column_to_rownames("ensembl")
  
  if (params$gene_sets_present == TRUE) {
    gene_set_names <- annotation_df %>%
      dplyr::arrange(gene_set) %>%
      dplyr::pull(gene_set) %>%
      unique()
    
    # get colors for the annotation object
    annotation_colors <-
      palette.colors(palette = "Okabe-Ito")[2:(length(gene_set_names) + 1)]
    
    gene_set_colors = annotation_colors[1:length(gene_set_names)]
    names(gene_set_colors) = gene_set_names
    gene_set_colors <- as.list(gene_set_colors)
    
    # create the column annotation for the ComplexHeatmap
    column_annotation <- HeatmapAnnotation(
      df = annotation_df,
      col = gene_set_colors,
      annotation_label = c("Gene Set")
    )
  }
}
```

The below heatmap displays genes on the x-axis, cells on the y-axis, and color indicates gene expression scaled using a z-score. 
Using only the provided marker genes, hierarchical clustering of cells was performed to identify if any clear structure or groupings of cells with similar gene expression patterns are present. 
The colored bar on the top of the graph indicates which gene set the provided marker gene corresponds to. 

```{r, fig.width=10, fig.height=10}
# create the ComplexHeatmap
Heatmap(normalized_sce_zscores_matrix,
        # implement a correlation based distance metric
        clustering_distance_rows = "pearson",
        # calculate the average distance between each cluster before merging
        clustering_method_rows = "average",
        clustering_distance_columns = "pearson",
        clustering_method_columns = "average",
        show_row_names = FALSE,
        show_column_names = TRUE,
        top_annotation = column_annotation,
        name = "z score")
```

## Single gene expression plots

### Sina plots

The plot below displays the normalized and transformed expression for each of the provided
marker genes in comparison to the mean expression of all genes present in the dataset. 

```{r}
# plot expression of marker genes
plot_markers_expression_sina(normalized_sce,
                             marker_genes,
                             params$ensembl_id_column,
                             params$gene_symbol_column)
```

### UMAP plots

Each plot below represents the UMAP results calculated using this sample's data.
Each dot represents a cell and the color indicates the individual marker gene expression (each plot being relevant to a single marker gene).

```{r}
plot_markers_expression_umap(normalized_sce,
                             marker_genes,
                             params$ensembl_id_column,
                             params$gene_symbol_column)
```

### PCA plots

Each plot below represents the PCA results calculated using this sample's data.
Each dot represents a cell and the color indicates the individual marker gene expression (each plot being relevant to a single marker gene).

```{r}
plot_markers_expression_pca(normalized_sce,
                            marker_genes,
                            params$ensembl_id_column,
                            params$gene_symbol_column)
```

## Session Info

```{r session_info}
sessionInfo()
```

