---
params:
  sample: Sample
  normalized_sce: "data/anderson-single-cell/normalized/normalized_GSM4186961_sce.rds"
  goi_list: "data/anderson-single-cell/goi-lists/mapped_goi_list.tsv"
  date: !r Sys.Date()
  gene_set_column: NULL
  ensembl_id_column: "ensembl"
  gene_symbol_column: NULL

title: "`r glue::glue('Gene expression plots for {params$sample}')`"
author: "CCDL"
date: "`r params$date`"
output: 
  html_document:
    toc: true
    toc_depth: 2
    toc_float: true
    number_sections: true
    code_folding: hide
---

## Set Up

```{r, setup, include=FALSE}
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE
)
```

```{r}
# Load libraries
library(magrittr)
library(readr)
library(SingleCellExperiment)
library(scater)
library(ggplot2)
library(ComplexHeatmap)

# Set seed for heatmap colors
set.seed(201)

# Source custom functions script
source(file.path("..", "utils", "provided-goi-analysis-functions.R"))
```

## Read in data

```{r}
normalized_sce <- read_rds(file.path(params$normalized_sce))

goi_list <- read_tsv(file.path(params$goi_list))

num_cells <- dim(normalized_sce)[2]
glue::glue("This sample has {num_cells} cells.")
```

## Perform Hierarchical Clustering

```{r}
ensembl_id_column_sym <- rlang::sym(params$ensembl_id_column)

# get logcounts from normalized sce object
normalized_sce_logcounts_matrix <- as.matrix(t(logcounts(normalized_sce)))

# filter counts matrix to only data associated with the provided genes of
# interest
normalized_sce_logcounts_matrix <- normalized_sce_logcounts_matrix[,colnames(normalized_sce_logcounts_matrix) %in% goi_list[[ensembl_id_column_sym]]]

# transform counts into z-scores
normalized_sce_zscores_matrix <- scale(normalized_sce_logcounts_matrix,
                                    center = TRUE, scale = TRUE)

# prepare the matrix and column annotation for heatmap plotting
if(!is.null(params$gene_set_column)) {
  if (!is.null(params$gene_symbol_column)) {
    normalized_sce_zscores_matrix <- colnames_to_gene_symbols(
        normalized_sce_zscores_matrix,
        goi_list,
        params$ensembl_id_column,
        params$gene_symbol_column
      )
    gene_id_column <- params$gene_symbol_column
    column_annotation <-
      prepare_heatmap_annotation(
        normalized_sce_zscores_matrix,
        goi_list,
        gene_id_column,
        params$gene_set_column
      )
  } else {
    gene_id_column <- params$ensembl_id_column
    column_annotation <-
      prepare_heatmap_annotation(
        normalized_sce_zscores_matrix,
        goi_list,
        gene_id_column,
        params$gene_set_column
      )
  }
} else {
  if (!is.null(params$gene_symbol_column)) {
    normalized_sce_zscores_matrix <-
      colnames_to_gene_symbols(
        normalized_sce_zscores_matrix,
        goi_list,
        params$ensembl_id_column,
        params$gene_symbol_column
      )
    column_annotation <- NULL
  } else {
    column_annotation <- NULL
  }
}
```

### Heatmap plotting

The below heatmap displays genes on the x-axis, cells on the y-axis, and color indicates gene expression scaled using a z-score. 
Using only the provided genes of interest, hierarchical clustering of cells was performed to identify if any clear structure or groupings of cells with similar gene expression patterns are present. 
The colored bar on the top of the graph indicates which gene set the provided gene of interest corresponds to. 

```{r, fig.width=10, fig.height=10}
# create the ComplexHeatmap
Heatmap(normalized_sce_zscores_matrix,
        # implement a correlation based distance metric
        clustering_distance_rows = "pearson",
        # calculate the average distance between each cluster before merging
        clustering_method_rows = "average",
        clustering_distance_columns = "pearson",
        clustering_method_columns = "average",
        show_row_names = FALSE,
        show_column_names = TRUE,
        top_annotation = column_annotation,
        name = "z score")
```

## Single gene expression plots

### Sina plots

The plot below displays the normalized and transformed expression for each of the provided
genes of interest in comparison to the mean expression of all genes present in the dataset. 

```{r}
# plot expression of genes of interest
plot_goi_expression_sina(normalized_sce,
                             goi_list,
                             params$ensembl_id_column,
                             params$gene_symbol_column)
```

### UMAP plots

Each plot below represents the UMAP results calculated using this sample's data.
Each dot represents a cell and the color indicates the individual gene of interest's expression (each plot being relevant to a single gene of interest).

```{r}
plot_goi_expression_umap(normalized_sce,
                             goi_list,
                             params$ensembl_id_column,
                             params$gene_symbol_column)
```

### PCA plots

Each plot below represents the PCA results calculated using this sample's data.
Each dot represents a cell and the color indicates the individual gene of interest's expression (each plot being relevant to a single gene of interest).

```{r}
plot_goi_expression_pca(normalized_sce,
                            goi_list,
                            params$ensembl_id_column,
                            params$gene_symbol_column)
```

## Session Info

```{r session_info}
sessionInfo()
```

